File: learning-core/executable/dbtx/cpp/dbtx.cpp
#include <stdexcept>
#include <string>
#include <unordered_map>

struct Store {
    std::unordered_map<std::string, int> balances;
};

struct Tx {
    std::unordered_map<std::string, int> working;
};

Tx begin(const Store& store) {
    return Tx{store.balances};
}

void transfer(Tx& tx, const std::string& source, const std::string& target, int amount) {
    if (amount <= 0) throw std::invalid_argument("amount must be positive");
    if (!tx.working.contains(source) || !tx.working.contains(target)) {
        throw std::out_of_range("account not found");
    }
    if (tx.working.at(source) < amount) {
        throw std::runtime_error("insufficient funds");
    }
    tx.working.at(source) -= amount;
    tx.working.at(target) += amount;
}

void commit(Store& store, Tx tx) {
    store.balances = std::move(tx.working);
}

File: learning-core/executable/dbtx/go/dbtx.go
package dbtx

import "errors"

var (
	ErrNotFound          = errors.New("account not found")
	ErrInsufficientFunds = errors.New("insufficient funds")
)

type Store struct {
	Balances map[string]int
}

type Tx struct {
	working map[string]int
}

func (s *Store) Begin() *Tx {
	copyMap := make(map[string]int, len(s.Balances))
	for k, v := range s.Balances {
		copyMap[k] = v
	}
	return &Tx{working: copyMap}
}

func (tx *Tx) Transfer(source, target string, amount int) error {
	if amount <= 0 {
		return errors.New("amount must be positive")
	}
	src, ok := tx.working[source]
	if !ok {
		return ErrNotFound
	}
	if _, ok := tx.working[target]; !ok {
		return ErrNotFound
	}
	if src < amount {
		return ErrInsufficientFunds
	}
	tx.working[source] -= amount
	tx.working[target] += amount
	return nil
}

func (s *Store) Commit(tx *Tx) { s.Balances = tx.working }
